File: TestRunnerApp/DevTests/PyPySe_encode.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;

/// <summary>
///     Experimental code for running python test scripts 
/// </summary>
/// 
/// <todo>
/// 
///     Workaround implemented: Receive exceptions raised in script    
///     DONE: Access selenium module from script
///     DONE: Access enums from script, workaround with switch statements for now
/// </todo>
/// 
/// <remarks>
///     Install IronPython from https://ironpython.net/  and edit path to Lib below
///     Install Selenium module using pip or easy_install, edit path to module below
///     Install-Package IronPython -Version 2.7.9
/// 
///     Copy both .dll and .py to Tests directory
/// </remarks>


namespace PythonTest
{
    // Python
    using Microsoft.Scripting;
    using Microsoft.Scripting.Hosting;
    using Microsoft.Scripting.Utils;
    using IronPython.Hosting;

    // TestRunnerApp
    using TestRunnerLib;

    // Selenium
    using OpenQA.Selenium;
    using OpenQA.Selenium.Support.UI;
    

    public class PyPySe_encode : IWebTest
    {
        public TestResult Test(WebDriverType webDriverType, string[] testData)
        {
            ScriptEngine engine = Python.CreateEngine();

            // Python paths
            ICollection<string> searchPaths = engine.GetSearchPaths();
            // Project root
            searchPaths.Add("..\\.."); // 
           // IronPython lib
            searchPaths.Add(@"C:\Program Files\IronPython 2.7\Lib");
            // Selenium module
            searchPaths.Add(@"C:\Program Files\IronPython 2.7\Lib\site-packages");
            //searchPaths.Add(@"C:\Users\joakim\AppData\Local\Programs\Python\Python37-32\Lib\site-packages");
            engine.SetSearchPaths(searchPaths);

            // Create python module that contains Outcome, WebDriverType, TestResult
            var testModuleSource = String.Join(
                Environment.NewLine,
                "import clr",
               $"clr.AddReferenceToFileAndPath(r'{typeof(Outcome).Assembly.Location}')",
               $"clr.AddReferenceToFileAndPath(r'{typeof(TestResult).Assembly.Location}')",
               $"clr.AddReferenceToFileAndPath(r'{typeof(WebDriverType).Assembly.Location}')",
               $"import {typeof(Outcome).FullName} as Outcome",
               $"import {typeof(TestResult).FullName} as TestResult",
               $"import {typeof(WebDriverType).FullName} as WebDriverType"
            );
            var testModule = engine.CreateModule("TestAppLib");
            engine.Execute(testModuleSource, testModule);

            // Scope for variables
            ScriptScope scope = engine.CreateScope();

            // Script
            ScriptSource script = engine.CreateScriptSourceFromFile("Tests\\PyPySe_encode.py");

            // Run variables, to be overwritten by script return
            int step = 0;
            string message = String.Empty;
            Outcome outcome = Outcome.NotRun;
            Exception rex = null;

            var los = new List<string>();
            foreach (string s in testData)
            {
                Debug.WriteLine("string: " + s);
                byte[] bytes = Encoding.Default.GetBytes(s);
                los.Add(Encoding.UTF7.GetString(bytes));
                Debug.WriteLine("handled string: " + los.Last());
            }
            testData = los.ToArray();

            //byte[] bytes = Encoding.Unicode.GetBytes(testData[2]);
            //testData[2] = Encoding.UTF8.GetString(bytes);


            /* Access from script */
            // Make sure we can get these later
            scope.SetVariable("outcome", outcome);
            scope.SetVariable("step", step);
            scope.SetVariable("message", message);
            scope.SetVariable("rex", rex);
            // Actual input params
            scope.SetVariable("webdrivertype", webDriverType);
            scope.SetVariable("testData", testData);


            

            try
            {
               script.Execute(scope);
               TestResult result = scope.GetVariable<TestResult>("result");

                // Work around any mistakenly returned Python exceptions
                result.e = null;

                // Exception data from Python exception string
                int exIndex = result.message.IndexOf("Exception:");
               if (exIndex > -1)
               {
                   result.eType = result.message.Substring(0, exIndex + 9);
                   result.eMessage = result.message.Substring(exIndex + 11);
                   result.message = "Exception";
               }

               return result;
            }

            // Web driver error
            catch (OpenQA.Selenium.DriverServiceNotFoundException ex)
            {
                return new TestResult(Outcome.Warning, step, ex);
            }
            // Selenium exception during action
            catch (OpenQA.Selenium.WebDriverException ex)
            {
                return new TestResult(Outcome.Fail, step, ex);
            }
            // Other errors
            catch (Exception ex)
            {
                return new TestResult(Outcome.Warning, step, ex);
            }


            
        }
    }
}

File: TestRunnerApp/DevTests/PyPySe_encode.py
# -- coding: utf-8 --
import time
import sys
import clr

from selenium import webdriver
from selenium.webdriver.support.ui import WebDriverWait
from selenium.common.exceptions import *
from selenium.webdriver.support import *
from selenium.webdriver.support import expected_conditions

from TestAppLib import TestResult, Outcome, WebDriverType

# Format exception string
def ExceptionFormat(ex):
    try:
        name = repr(ex).replace("()", "")
        msg = str(ex).replace("Message", "")    
        return name + msg
    except Exception:
        return "Error formatting exception"

# Get webdriver from UI selection
def GetDriver(webdrivertype):
    try:
      if webdrivertype == WebDriverType.Chrome:
          return webdriver.Chrome()
      elif webdrivertype == WebDriverType.Firefox:
          return webdriver.Firefox()
      elif webdrivertype == WebDriverType.IE:
          return webdriver.Ie()
      # Not found
      raise RemoteDriverServerException

    except Exception:
        result = TestResult(Outcome.Warning, 0, "WebDriver error")

# Test runner
def Run(driver, testData):
      server = testData[1]
      #server = testData[1].encode("utf7", "strict");
      #input = testData[2].encode("utf7", "strict");
      #input = input.replace("+APY-", "ö");
      #input = testData[2]

      input = "ö".encode('UTF-8')

      step = 1
      try:
          driver.get(server)
          el = driver.find_element_by_name("q")
          step = 2
          el.send_keys(input)
          step = 3
          el.submit()
          step = 4
          #WebDriverWait(driver, 10).until(expected_conditions.title_contains(input))
          WebDriverWait(driver, 20)
          step = 5
          return TestResult(Outcome.Pass, "Page title is: {title}".format(title=driver.title) + "Funny letter: " + input)

       # Selenium exception during action
      except WebDriverException as ex:
           #s = repr(ex) + str(ex);
           return TestResult(Outcome.Fail, step, ExceptionFormat(ex))
       
       # Other errors
      except Exception as ex:
           return TestResult(Outcome.Fail, step, ExceptionFormat(ex))

      finally:
             driver.quit()

# Execution
#reload(sys)  
#sys.setdefaultencoding('utf8')
result = Run(GetDriver(webdrivertype), testData)
